File: patterns/debouncer.swift
//
//  FunctionDebounce.swift
//  Rise
//
//  Created by Peter Tong on 22/11/2024.
//

import Foundation
import UIKit

/// A generic class to debounce function calls.
///
/// **Purpose**:
/// - Debouncing is used to delay the execution of a function until a certain time has elapsed since the last time it was invoked.
/// - This is useful for costly operations (e.g., API calls, file uploads) where frequent calls are unnecessary, and the last call is the most important.
///
/// **Features**:
/// 1. Supports asynchronous functions (`async`/`await`).
/// 2. Automatically cancels pending tasks when the app goes to the background or exits.
/// 3. Singleton instance for easy global access.
/// 4. Thread-safe implementation using GCD.
///
/// **Usage**:
/// ```swift
/// Debouncer.shared.debounceCall(delay: 2.0) {
///     await uploadDataToCloud() // Debounced async function
/// }
/// ```
public class Debouncer {
    
    /// Singleton instance for global access
    public static let shared = Debouncer()
    
    /// A dictionary to hold pending tasks for different identifiers
    private var pendingTasks: [String: Task<Void, Never>] = [:]
    
    /// Serial queue to ensure thread safety
    private let queue = DispatchQueue(label: "com.debouncer.queue")

    /// Private initializer to enforce singleton usage
    private init() {}

    deinit {}

    /// Debounce a function call with a specified delay.
    ///
    /// - Parameters:
    ///   - id: A unique identifier for the debounced task (default is a random UUID).
    ///   - delay: The time in seconds to wait before executing the function.
    ///   - task: The function to be executed (can be `async`).
    public func debounceCall(
        id: String = UUID().uuidString,
        delay: TimeInterval,
        task: @escaping @Sendable () async -> Void
    ) {
        // Ensure thread safety when accessing `pendingTasks`
        queue.sync {
            // Cancel the existing task for this ID if it exists
            if let existingTask = pendingTasks[id] {
                existingTask.cancel()
            }
            
            // Create a new task to be executed after the delay
            let newTask = Task {
                try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
                if !Task.isCancelled {
                    await task()
                }
            }
            
            // Store the new task in the dictionary
            pendingTasks[id] = newTask
        }
    }

    /// Cancel all pending tasks (e.g., when the app goes to the background).
    @objc private func cancelAllPendingTasks() {
        queue.sync {
            for (_, task) in pendingTasks {
                task.cancel()
            }
            pendingTasks.removeAll()
        }
    }
}

File: patterns/debouncer.kt
package build.designand.riselife.patterns

// Debouncer.kt

import kotlinx.coroutines.*
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import kotlin.time.Duration
import kotlin.time.DurationUnit
import kotlin.time.toDuration

/**
 * A singleton class to debounce function calls.
 *
 * **Purpose**:
 * - Debouncing delays the execution of a function until a specified time has elapsed since the last invocation.
 * - Useful for preventing excessive calls to costly operations.
 *
 * **Features**:
 * 1. Supports suspending functions.
 * 2. Automatically cancels pending tasks when [cancelAllPendingTasks] is called.
 * 3. Thread-safe implementation using [ConcurrentHashMap].
 *
 * **Usage**:
 * ```kotlin
 * Debouncer.debounceCall(delay = 2.seconds) {
 *     uploadDataToCloud() // Debounced suspending function
 * }
 * ```
 */
object Debouncer {

    // CoroutineScope for managing debounce jobs.
    // Using SupervisorJob to prevent failure of one job from cancelling others.
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    // Map to hold pending debounce jobs identified by unique IDs.
    private val pendingJobs = ConcurrentHashMap<String, Job>()

    /**
     * Debounce a function call with a specified delay.
     *
     * @param id A unique identifier for the debounced task. Defaults to a random UUID.
     * @param delay The delay duration before executing the task.
     * @param task The suspending function to execute after the delay.
     */
    fun debounceCall(
        id: String = UUID.randomUUID().toString(),
        delay: Duration,
        task: suspend () -> Unit
    ) {
        // Cancel any existing job with the same ID.
        pendingJobs[id]?.cancel()

        // Launch a new coroutine for the debounced task.
        val job = scope.launch {
            delay(delay.inWholeMilliseconds)
            try {
                task()
            } catch (e: CancellationException) {
                // Task was cancelled; no action needed.
                // Optionally, log if needed.
            }
        }

        // Store the new job in the pendingJobs map.
        pendingJobs[id] = job
    }

    /**
     * Cancel all pending debounce tasks.
     * Call this method when the app goes to the background or exits.
     */
    fun cancelAllPendingTasks() {
        for ((id, job) in pendingJobs) {
            job.cancel()
        }
        pendingJobs.clear()
    }

    /**
     * Cancel a specific pending debounce task by its ID.
     *
     * @param id The unique identifier of the task to cancel.
     */
    fun cancelTask(id: String) {
        pendingJobs[id]?.cancel()
        pendingJobs.remove(id)
    }
}
